### PR TITLE
Feature/add onnxrt openvino hardware id parameter

### PR DESCRIPTION
**Pull request by jafet-chaves**
Changes summary:

* Add parameter class (and tests) for ONNXRT OpenVINO backend. Meta information is refactored from the base class, most of the methods can not be reused because the base class implementation is based on a friend class Accesor member and friendship can not be inherited in C++. This limits greatly the possibility to refactor code and instead of rewriting the base class I opted for overrriding the methods in the derived class.
* Change default value of intra_num_threads property in ONNXRT engine, from 1 to 0. I feel this change is preferable since for ONNXRT (and execution providers) there is a better performance when this property is to 0 instead of 1.

### CODE REVIEW

File: r2i/onnxrt/engine.cc [hunk 1]
```diff
@@ -28,7 +28,7 @@ namespace onnxrt {
 Engine::Engine () {
   /* Initialize all variable members */
   logging_level = OrtLoggingLevel::ORT_LOGGING_LEVEL_WARNING;
-  intra_num_threads = 1;
+  intra_num_threads = 0;
```
Comment (lumurillo) [inline r2i/onnxrt/engine.cc:31]: Just to understand, what is the reason for this change?

Comment (jafet-chaves) [reply r2i/onnxrt/engine.cc:31]: From my tests I constantly encounter that a better performance is obtained when this property is set to 0, I believe this optimization should be by default.

File: r2i/onnxrt_openvino/parameters.cc [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+  if (nullptr == in_model) {
+    error.Set (RuntimeError::Code::NULL_PARAMETER, "Received null model");
+    return error;
+  }
+
+  auto engine = std::dynamic_pointer_cast<Engine, IEngine>(in_engine);
+  if (nullptr == engine) {
+    error.Set (RuntimeError::Code::INCOMPATIBLE_ENGINE,
+               "The provided engine is not an onnxrt engine");
```
Comment (emmadrigal) [inline r2i/onnxrt_openvino/parameters.cc:71]: Is the standard to use `onnxrt` in lower-case, in the documentation it apperas as `ONNXRT`

File: r2i/onnxrt_openvino/parameters.cc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    return error;
+  }
+
+  value = accessor->value;
+  return error;
+}
+
+RuntimeError Parameters::Get (const std::string &in_parameter, double &value) {
+  RuntimeError error;
+  return error;
```
Comment (emmadrigal) [inline r2i/onnxrt_openvino/parameters.cc:142]: Why isn't this implemented?

Shouldn't we report an error if someone tries to use this call?

File: r2i/onnxrt_openvino/parameters.h [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  friend class Accessor;
+
+  class Accessor {
+   public:
+    Accessor (Parameters *target) : target(target) {}
+    virtual RuntimeError Set () = 0;
+    virtual RuntimeError Get () = 0;
+    virtual ~Accessor() {}
+
+   public:
```
Comment (emmadrigal) [inline r2i/onnxrt_openvino/parameters.h:55]: Why do we need public twice in this class?

File: r2i/onnxrt/engine.cc (hunk 1 above)
Comment (emmadrigal) [inline r2i/onnxrt/engine.cc:31]: Since you found this value empirically, add a comment to explain why this is `0`

File: r2i/onnxrt_openvino/parameters.h (hunk 4 above)
Comment (jafet-chaves) [reply r2i/onnxrt_openvino/parameters.h:55]: We don't. I just noticed this is in all the other backends aswell. It doesn't affect anything but look ugly.

File: r2i/onnxrt_openvino/parameters.cc (hunk 2 above)
Comment (jafet-chaves) [reply r2i/onnxrt_openvino/parameters.cc:71]: I agree on using ONNXRT instead

File: r2i/onnxrt_openvino/parameters.cc (hunk 3 above)
Comment (jafet-chaves) [reply r2i/onnxrt_openvino/parameters.cc:142]: I agree, I'll return RuntimeError::Code::NOT_IMPLEMENTED in the Get() and Set() methods.

File: r2i/onnxrt_openvino/parameters.cc [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+  ParamDesc hardware_id_desc = {
+    {hardware_id_meta},
+    std::make_shared<HardwareIdAccessor>(this)
+  };
+  parameter_map.emplace(std::make_pair(hardware_id_meta.name, hardware_id_desc));
+}
+
+RuntimeError Parameters::Configure (std::shared_ptr<r2i::IEngine> in_engine,
+                                    std::shared_ptr<r2i::IModel> in_model) {
```
Comment (mmontero) [inline r2i/onnxrt_openvino/parameters.cc:55]: here you have 2 different types of indentation aligning the second parameters line to the opening ( and 4 spaces indentation, I don't know what is the standard you are following but make it consistent

File: r2i/onnxrt_openvino/parameters.cc [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    return error;
+  }
+
+  value = accessor->value;
+  return error;
+}
+
+RuntimeError Parameters::Get (const std::string &in_parameter, double &value) {
+  RuntimeError error;
+  return error;
```
Comment (mmontero) [inline r2i/onnxrt_openvino/parameters.cc:142]: shouldn't you be setting some kind of error here?

File: tests/unit/r2i/onnxrt_openvino/Makefile.am [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+# RidgeRun, LLC.  The user is free to modify the source code after obtaining
+# a software license from RidgeRun.  All source code changes must be provided
+# back to RidgeRun without any encumbrance.
+
+AM_DEFAULT_SOURCE_EXT = .cc
+
+if ENABLE_TESTS
+if HAVE_ONNXRT_OPENVINO
+
+TESTS =                   \
```
Comment (mmontero) [inline tests/unit/r2i/onnxrt_openvino/Makefile.am:16]: this looks pretty weird move parameters to this line

File: r2i/onnxrt_openvino/parameters.cc [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+#include "parameters.h"
+
+#include <string>
+
+namespace r2i {
+namespace onnxrt_openvino {
+
+Parameters::Parameters () {
+  ParamDesc logging_level_desc = {
```
Comment (michaelgruner) [inline r2i/onnxrt_openvino/parameters.cc:20]: This is duplicated code. You have the exact same in the parent constructor.

File: r2i/onnxrt_openvino/parameters.cc [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+  this->model = model;
+
+  return error;
+}
+
+std::shared_ptr<r2i::IEngine> Parameters::GetEngine () {
+  return this->engine;
+}
+
+Parameters::ParamDesc Parameters::Validate (const std::string &in_parameter,
```
Comment (michaelgruner) [inline r2i/onnxrt_openvino/parameters.cc:92]: This is done in the parent already.

File: r2i/onnxrt_openvino/parameters.cc [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+  if (param.meta.type != type) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Parameter \"" + in_parameter + "\" is not of type " + stype);
+    return undefined;
+  }
+
+  return param;
+}
+
+RuntimeError Parameters::Get (const std::string &in_parameter, int &value) {
```
Comment (michaelgruner) [inline r2i/onnxrt_openvino/parameters.cc:119]: This is already done in the parent

File: r2i/onnxrt_openvino/parameters.h [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    Accessor (Parameters *target) : target(target) {}
+    virtual RuntimeError Set () = 0;
+    virtual RuntimeError Get () = 0;
+    virtual ~Accessor() {}
+
+   protected:
+    Parameters *target;
+  };
+
+  class StringAccessor : public Accessor {
```
Comment (michaelgruner) [inline r2i/onnxrt_openvino/parameters.h:59]: Why are all this here and we are not making use of the ones in the parent?

Comment (jafet-chaves) [reply r2i/onnxrt_openvino/parameters.h:59]: Answering this one and all related comments above. This is because you [can not inherite friendship in C++](https://stackoverflow.com/questions/3561648/why-does-c-not-allow-inherited-friendship#:~:text=A%20derived%20class%20can%20inherit,'member'%20of%20the%20base.&text=%2411.4%20%2D%20%22Also%2C%20because%20the,from%20the%20class%20granting%20friendship.%22), so basically nothing that uses the Accesor class in derived classes.

File: r2i/onnxrt_openvino/parameters.cc (hunk 6 above)
Comment (jafet-chaves) [reply r2i/onnxrt_openvino/parameters.cc:142]: Yes, we should but the error that I intend to return here (RuntimeError::Code::NOT_IMPLEMENTED), is defined in another branch: feature/add-pre-post-to-loader. I'll keep this change in mind once these changes have been merged.
